### PR TITLE
GraphQL Instrumentation

### PR DESCRIPTION
**Pull request by twcrone**
### Overview
GraphQL Instrumentation

### Related Github Issue
For every metric created, the agent creates a scoped and unscoped version. An unscoped metric could get filtered out from being sent by the agent if it doesn't meet certain criteria. This prevents the GraphQL metrics from being filtered.

### Checks
[X] Are your contributions backwards compatible with relevant frameworks and APIs?
[] Does your code contain any breaking changes? Please describe. 
[X] Does your code introduce any new dependencies? Please describe.

### CODE REVIEW

**Comment by XiXiaPdx** [comment]
Manual testing of NetFlix DGS example jav app  - looks good. https://staging.onenr.io/0mMRNAKLAwn

**Comment by twcrone** [comment]
So not really 50+ files of code.  We have our GraphQL test files in separate *.gql files under `src/test/resources`

**Comment by twcrone** [comment]
We are totally fine squashing or whatever team wants to do.

File: instrumentation/graphql-java-16.2/src/test/java/com/nr/instrumentation/graphql/GraphQLObfuscatorTest.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    private final static String OBFUSCATE_DATA_DIR = "obfuscateQueryTestData";
+
+    @ParameterizedTest
+    @CsvFileSource(resources = "/obfuscateQueryTestData/obfuscate-query-test-data.csv", delimiter = '|', numLinesToSkip = 2)
+    public void testObfuscateQuery(String queryToObfuscateFile, String expectedObfuscatedQueryFile) {
+        //setup
+        queryToObfuscateFile = queryToObfuscateFile.trim();
+        expectedObfuscatedQueryFile = expectedObfuscatedQueryFile.trim();
+        String expectedObfuscatedResult = readText(OBFUSCATE_DATA_DIR, expectedObfuscatedQueryFile);//readText(expectedObfuscatedQueryFile);
```
Comment (jasonjkeller) [inline instrumentation/graphql-java-16.2/src/test/java/com/nr/instrumentation/graphql/GraphQLObfuscatorTest.java:19]: Can this commented out code (`//readText(expectedObfuscatedQueryFile);`) be removed?

**Review by jasonjkeller — APPROVED** [review]
A few minor nits but overall looks good!

File: instrumentation/graphql-java-16.2/src/test/java/com/nr/instrumentation/graphql/GraphQLObfuscatorTest.java (hunk 1 above)
Comment (twcrone) [reply instrumentation/graphql-java-16.2/src/test/java/com/nr/instrumentation/graphql/GraphQLObfuscatorTest.java:19]: Yes, definitely...good catch

File: newrelic-agent/src/main/java/com/newrelic/agent/service/analytics/TracerToSpanEvent.java [hunk 2]
```diff
@@ -122,6 +123,17 @@ public SpanEvent createSpanEvent(Tracer tracer, TransactionData transactionData,
         return builder.build();
     }
 
+    private SpanEventFactory maybeSetGraphQLAttributes(Tracer tracer, SpanEventFactory builder) {
+        Map<String, Object> agentAttributes = tracer.getAgentAttributes();
+        boolean containsGraphQLAttributes = agentAttributes.keySet().stream().anyMatch(key -> key.contains("graphql"));
+        if(containsGraphQLAttributes){
```
[the remark this answers is not in the document]
Comment (twcrone) [reply newrelic-agent/src/main/java/com/newrelic/agent/service/analytics/TracerToSpanEvent.java:129]: okey, dokey

**Review by jasonjkeller — APPROVED** [review]

File: instrumentation/graphql-java-16.2/build.gradle [hunk 3]
```diff
@@ -0,0 +1,31 @@
+dependencies {
+    implementation(project(":agent-bridge"))
+
+    implementation 'com.graphql-java:graphql-java:16.2'
```
Comment (meiao) [inline instrumentation/graphql-java-16.2/build.gradle:4]: Do we need to add something to the `THIRD_PARTY_NOTICES.md` file?

**Review by meiao — COMMENTED** [review]
Mostly nit picking.

File: instrumentation/graphql-java-16.2/src/main/java/com/nr/instrumentation/graphql/GraphQLSpanUtil.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    private static void setDefaultOperationAttributes(String query) {
+        AgentBridge.privateApi.addTracerParameter("graphql.operation.type", DEFAULT_OPERATION_TYPE);
+        AgentBridge.privateApi.addTracerParameter("graphql.operation.name", DEFAULT_OPERATION_NAME);
+        AgentBridge.privateApi.addTracerParameter("graphql.operation.query", obfuscate(query));
+    }
+
+    private static <T> T getValueOrDefault(T value, T defaultValue) {
+        return value == null ? defaultValue : value;
+    }
```
[the remark this answers is not in the document]
Comment (twcrone) [reply instrumentation/graphql-java-16.2/src/main/java/com/nr/instrumentation/graphql/GraphQLSpanUtil.java:60]: Yeah, I don't know.  I'm not sure where something like this goes.

File: instrumentation/graphql-java-16.2/src/main/java/com/nr/instrumentation/graphql/GraphQLTransactionName.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        return onlyNonFederatedSelectionOrNoneFrom(selectionSet);
+    }
+
+    private static boolean notFederatedFieldName(final String fieldName) {
+        return !(TYPENAME.equals(fieldName) || ID.equals(fieldName));
+    }
+
+    private static boolean isNullOrEmpty(final Collection<?> c) {
+        return c == null || c.isEmpty();
+    }
```
[the remark this answers is not in the document]
Comment (twcrone) [reply instrumentation/graphql-java-16.2/src/main/java/com/nr/instrumentation/graphql/GraphQLTransactionName.java:152]: My rule is, if you use it twice, copy paste is fine.  But if you go to use it in a third place, pull it out.  But I can pull it out.

File: instrumentation/graphql-java-16.2/build.gradle (hunk 3 above)
Comment (twcrone) [reply instrumentation/graphql-java-16.2/build.gradle:4]: ??? this is the library we are instrumenting.  I don't know about this type of file.

File: instrumentation/graphql-java-16.2/src/main/java/com/nr/instrumentation/graphql/GraphQLTransactionName.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (twcrone) [reply instrumentation/graphql-java-16.2/src/main/java/com/nr/instrumentation/graphql/GraphQLTransactionName.java:152]: Typically I use a Collection Utils class that exists in latter versions of Java

File: instrumentation/graphql-java-16.2/build.gradle (hunk 3 above)
Comment (jasonjkeller) [reply instrumentation/graphql-java-16.2/build.gradle:4]: Not for dependencies used in instrumentation modules.

File: instrumentation/graphql-java-16.2/src/main/java/com/nr/instrumentation/graphql/GraphQLErrorHandler.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import graphql.execution.FieldValueInfo;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.logging.Level;
+
+public class GraphQLErrorHandler {
+    public static void reportResolverThrowableToNR(Throwable e) {
+        NewRelic.noticeError(e);
+    }
```
[the remark this answers is not in the document]
Comment (twcrone) [reply instrumentation/graphql-java-16.2/src/main/java/com/nr/instrumentation/graphql/GraphQLErrorHandler.java:24]: Ok

File: instrumentation/graphql-java-16.2/build.gradle (hunk 3 above)
Comment (jasonjkeller) [reply instrumentation/graphql-java-16.2/build.gradle:4]: Typically it's just libraries that we directly use in development of the agent that need to be declared.

**Review by meiao — APPROVED** [review]

**Pull request merged**